ReactWindows/ReactNative.Net46/Bridge/DispatcherHelpers.cs

```diff
@@ -151,20 +151,82 @@ public static Task<T> CallOnDispatcher<T>(Dispatcher dispatcher, Func<T> func, b
         {
             if (allowInlining && IsOnDispatcher(dispatcher))
             {
-                return Task.FromResult(func());
+                try
+                {
+                    var result = func();
+                    return Task.FromResult(result);
+                }
+                catch (Exception e)
+                {
+                    return Task.FromException<T>(e);
+                }
             }
-            else
-            {
-                var taskCompletionSource = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
 
-                RunOnDispatcher(dispatcher, DispatcherPriority.Normal, () =>
+            var taskCompletionSource = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            RunOnDispatcher(dispatcher, DispatcherPriority.Normal, () =>
+            {
+                try
                 {
                     var result = func();
                     taskCompletionSource.SetResult(result);
+                }
+                catch (Exception e)
+                {
+                    taskCompletionSource.SetException(e);
+                }
+            });
+
+            return taskCompletionSource.Task;
+        }
+
+        /// <summary>
+        /// Invokes an async function on a specified dispatcher and asynchronously returns the
+        /// result.
+        /// </summary>
+        /// <param name="dispatcher">The Dispatcher to be used.</param>
+        /// <param name="func">The function to invoke.</param>
+        /// <param name="allowInlining">True if inlining is allowed when calling thread is on the same dispatcher as the one in the parameter.</param>
+        /// <returns>A task to await the result.</returns>
+        public static Task CallOnDispatcher(Dispatcher dispatcher, Func<Task> func, bool allowInlining = false)
+        {
+            if (allowInlining && IsOnDispatcher(dispatcher))
+            {
+                var result = Task.Run(async () =>
+                {
+                    try
+                    {
+                        await func();
+                        return Task.FromResult(true);
```

Comment:
I thought the reason of this new method to exist was the impossibility of doing CallOnDispatcher<void>.
But the implementation seems very different. Basically func() is called on a thread pool thread (Task.Run does that) rather than on the dispatcher that was passed as param, or through a straight call in this branch (since inlining is allowed and we are on the right dispatcher)